Throttling enhancements.

cstamas opened the pull request:
Implemented Mike's comments. This time, he _did not_ create an issue for these :P
- exposed all throttling variables with fixed names
- implemented lower limiting number sequence
- added possibility to "tune" minimum sleep time
- Added total runtime to log when upgrade ends

cstamas committed:
Throttling enhancements.

* exposed all throttling variables with fixed names
* implemented lower limiting number sequence
* added possibility to "tune" minimum sleep time
* Added total runtime to log when upgrade ends

velo commented on nexus/nexus-utils/src/main/java/org/sonatype/nexus/util/LowerLimitNumberSequence.java:
```diff
@@ -0,0 +1,75 @@
+package org.sonatype.nexus.util;
```
License header and UT missing

cstamas replied:
Fixed.

velo commented on nexus/nexus-utils/src/main/java/org/sonatype/nexus/util/NumberSequenceWrapper.java:
```diff
@@ -0,0 +1,49 @@
+package org.sonatype.nexus.util;
```
License header and Junit

cstamas replied:
License headers fixed, and for the other... "eh?" :)

cstamas committed:
Added source headers.

cstamas committed:
A modest UT

mah0526 commented on nexus/nexus-utils/src/main/java/org/sonatype/nexus/util/LowerLimitNumberSequence.java:
```diff
@@ -0,0 +1,87 @@
+        else
+        {
+            // perform the state change
+            final long wrappedPrev = super.prev();
```
EDIT:  I see the error in my ways.  the prior conditional check handles this correctly, though you will still sit at the lower limit for one extra call to next(), which is trivial.

This looks like it will move the sequence to a lower number, even if you are already at the low point and will continue to do so potentially indefinitely.  The implication seems to be (if I have this right), then you could have to advance forward potentially many times before you get up above the lowerLimit value -- i.e. you could return the value of lowerLimit repeatedly following calls to .next(), kind of like a Toyota Prius with the gas pedal stuck to the floor :)

cstamas committed:
Moving method to proper place

cstamas committed:
Attribute fix.

Both maps has to have key removed, as the "legacy flag" gets into
defaults map on overlay... and that is wrong (ie, remove will not work
once overlaid!). This is fixed by this commit.

nabcos commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/attributes/internal/DefaultAttributes.java:
```diff
@@ -73,7 +73,18 @@ public String put( String key, String value )
+        // key needs to be removed from both maps
+        final String valuesPrev = values.remove( key );
+        final String defaultsPrev = defaults.remove( key );
+        // but what we "saw" before depends where was the removed value
```
needs a brief description of what was seen before IMO (or issue ref, ...)

cstamas replied:
In short (I spotted this on RSO), the "marker" key was left in and persisted:
https://github.com/sonatype/nexus/blob/master/nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/TransitioningAttributeStorage.java#L37

Because, item is "lifted", then "attributes", and attributes are _overlaid_ to the item's attributes... meaning, they went in into "defaults" map.

And the transitioning attr storage did call "remove" but it was actually not removing it...
